// casti/entities/AddOn.js
require('shelljs/global')
const {Success, Failure} = require('../libs/types')

/**
 * 
 */

let removeAddOn = ({idOrName, organization}) => {
  try {
    let cmd = `clever addon delete ${idOrName} --org ${organization} --yes; `
    let result = exec(cmd)
    if(result.code !== 0) { return Failure.of(result.stderr) }
    return Success.of(idOrName)
  } catch (error) {
    return Failure.of(error.message)
  }
} 

class AddOn {
  constructor({type, name, organization, plan, region, id=null}) {
    this.type = type
    this.name = name
    this.organization = organization
    this.plan = plan
    this.region = region
    this.id = id
  }
  static of({type, name, organization, plan, region, id=null}) {
    return new this({type, name, organization, plan, region, id})
  }

  create() {
    try {
      let cmd = `clever addon create ${this.type.name} "${this.name}" --plan ${this.plan} --org ${this.organization} --region ${this.region}; `
      let result = exec(cmd)
      if(result.code !== 0) { return Failure.of(result.stderr) }

      //console.log(this.type)
  
      return Success.of({out: result.stdout, addon: {name: this.name, type: this.type}})  

    } catch (error) {
      return Failure.of(error.message)
    }

  }

  removeByName() {
    return removeAddOn({idOrName: this.name, organization: this.organization})
  }

  removeById() {
    // TODO
    // WIP 🚧: I need to find a way to get the id of the addon
  }


}

module.exports = AddOn
